src/Network/Wai/SAML2/Request.hs

```diff
@@ -0,0 +1,53 @@
+module Network.Wai.SAML2.Request
+    ( Request(..)
+    , issueRequest
+    ) where
+
+import Crypto.Random
+import Data.Time.Clock
+import Data.Time.Format
+import Network.Wai.SAML2.XML
+import Text.XML
+import qualified Data.ByteString as B
+import qualified Data.ByteString.Base64 as Base64
+import qualified Data.ByteString.Lazy as BL
+import qualified Data.Map.Strict as Map
+import qualified Data.Text as T
+import qualified Data.Text.Encoding as T
+
+-- | Parameters for SP-initiated SSO
+data Request = Request
+  { requestIssuer :: T.Text
+  , requestAllowCreate :: Bool
+  , requestNameIDFormat :: T.Text
+  }
+
+-- | Generated a base64-encoded AuthnRequest for SP initiated SSO.
+-- See also: http://docs.oasis-open.org/security/saml/Post2.0/sstc-saml-tech-overview-2.0-cd-02.html#5.1.2.SP-Initiated%20SSO:%20%20Redirect/POST%20Bindings|outline
+issueRequest :: Request -> IO B.ByteString
+issueRequest Request{..} = do
+    now <- getCurrentTime
+    let timestamp = T.pack $ formatTime defaultTimeLocale "%Y-%m-%dT%H:%M:%SZ" now
+    ident <- T.decodeUtf8 . Base64.encode <$> getRandomBytes 64
```

Comment:
According to [the SAML2 specification](https://docs.oasis-open.org/security/saml/v2.0/saml-core-2.0-os.pdf) (Section 3.2.1, Line 1467 onward), we must check that the identifier in the response matches the identifier in the request. I don't see a way to do that with the current implementation, since `ident` isn't returned to the caller in an easily accessible way (or provided by the caller).

Reply:
That's a good point. I split the function into two; `issueAuthnRequest :: Text -> IO AuthnRequest` generates a default `AuthnRequest` along with a timestamp and ID, and `renderAuthnRequest :: AuthnRequest -> ByteString`. This also makes the datatype correspond better with AuthnRequest on specification.

Reply:
Thanks, that seems like a better way to organise this.